File: backend/dataset/coinChange/coinChange_15.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        self.ans = float('inf')
		
		# Start searching from the biggest coin
        coins.sort(reverse=True)
        self.dfs(coins, amount, 0)
        return -1 if self.ans == float('inf') else self.ans

    
    def dfs(self, coins, amount, prev_count):
        """
        Recursive DFS function to seach valid coins combination.
        First is to use greedy method find out a potential-best solution.
        Then start to search the second biggest coin with pruning when the coins number >= the potential-best solution.

        Args:
            coins: coins list from which we pick coins into combination
            amount: target amount 
            prev_count: number of coins picked before this round
        
        """
        # Set up stop condtion
        if len(coins) == 0:
            return		

        if amount % coins[0] == 0:
			# Update potential answer
            self.ans = min(self.ans, prev_count + amount // coins[0])
        else:
            for k in range(amount // coins[0], -1, -1):
				# Set up pruning condtion
                if prev_count + k >= self.ans: 
                    break
                self.dfs(coins[1:], amount - k * coins[0], prev_count + k)

File: backend/CDataset/coinChange/coinChange_15.cpp
/*
dp[amount] = dp[amount-coins[i]] + 1;
*/

class Solution {
public:
    int coinChange(vector<int>& nums, int amount) {
        long long int n=nums.size();
        sort(nums.begin(),nums.end());
        //dp[i] -> min no of coins required to make amount 'i'
        vector<long long int> dp(amount+1,INT_MAX);
        dp[0]=0;
        for(long long int i=0;i<n;i++){
            if(nums[i]>amount)
                break;
            dp[nums[i]]=1;
        }
        
        for(long long int i=1;i<=amount;i++){
            if(dp[i]!=INT_MAX){
                for(long long int j=0;j<n && i+nums[j]<=amount;j++){
                    int new_val=i+nums[j];
                    if(new_val>amount)
                        break;
                    dp[new_val]=min(dp[new_val],dp[i]+1);
                }
            }
            
        }
        if(dp[amount]==INT_MAX){
            return -1;
        }
        return dp[amount];
    }
};
